gitlab-event/handler.go

```diff
@@ -0,0 +1,410 @@
+	}
+	digest := hmac.Sign(req, []byte(payloadSecret))
+	pushReq.Header.Add(sdk.CloudSignatureHeader, "sha1="+hex.EncodeToString(digest))
+
+	for k, v := range headers {
+		pushReq.Header.Add(k, v)
+	}
+
+	res, err := c.Do(pushReq)
+	if err != nil {
+		msg := "cannot post to " + function + ": " + err.Error()
+		auditEvent := sdk.AuditEvent{
+			Message: msg,
+			Source:  Source,
+		}
+		sdk.PostAudit(auditEvent)
+		return "", http.StatusInternalServerError, fmt.Errorf(msg)
+	}
+
+	if res.Body != nil {
+		defer res.Body.Close()
+	}
+	body, bodyErr := ioutil.ReadAll(res.Body)
+	if bodyErr != nil {
+		return "", http.StatusInternalServerError, bodyErr
+	}
+
+	if res.StatusCode != http.StatusOK {
+		err = fmt.Errorf(string(body))
+	}
+
+	return string(body), res.StatusCode, err
+}
+
+func InstalledApp(id int, instance, apiToken, installationTag string) (bool, error) {
+	wholeURL := instance + "/api/v4/projects/" + strconv.Itoa(id)
+
+	req, reqErr := http.NewRequest(http.MethodGet, wholeURL, nil)
+	if reqErr != nil {
+		return false, reqErr
+	}
+	req.Header.Add("PRIVATE-TOKEN", apiToken)
+
+	client := &http.Client{}
+	resp, respErr := client.Do(req)
+	if respErr != nil {
+		return false, respErr
+	}
+	defer resp.Body.Close()
+
+	body, bodyErr := ioutil.ReadAll(resp.Body)
+	if bodyErr != nil {
+		return false, bodyErr
+	}
+	projectInfo := GitLabProjectTags{}
+
+	unmarshalErr := json.Unmarshal(body, &projectInfo)
+	if unmarshalErr != nil {
+		return false, unmarshalErr
+	}
+
+	for _, tag := range projectInfo.TagList {
+		if strings.EqualFold(tag, installationTag) {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+func readBool(key string) bool {
+	if val, exists := os.LookupEnv(key); exists {
+		return val == "true" || val == "1"
+	}
+	return false
+}
+
+func tokenMatch(gitlabToken string, token string) bool {
+	return gitlabToken == token
+}
+
+func getCustomers(customerURL string) ([]string, error) {
+	customers := []string{}
+	if len(customerURL) == 0 {
+		return nil, fmt.Errorf("customerURL was nil")
+	}
+	c := http.Client{}
+	httpReq, httpErr := http.NewRequest(http.MethodGet, customerURL, nil)
+	if httpErr != nil {
+		return customers, httpErr
+	}
+	res, reqErr := c.Do(httpReq)
+	if reqErr != nil {
+		return customers, reqErr
+	}
+	if res.Body != nil {
+		defer res.Body.Close()
+		pageBody, bodyErr := ioutil.ReadAll(res.Body)
+		if bodyErr != nil {
+			return customers, bodyErr
+		}
+		customers = strings.Split(string(pageBody), "\n")
+		for i, c := range customers {
+			customers[i] = strings.ToLower(c)
+			customers[i] = strings.TrimSuffix(c, "\r")
+		}
+	}
+	return customers, nil
+}
+
+func validCustomer(customers []string, owner string) bool {
+	found := false
+	for _, customer := range customers {
+		if len(customer) > 0 &&
+			strings.EqualFold(customer, owner) {
+			found = true
+			break
+		}
+	}
+	return found
+}
+
+func getUser(pathWithNamespace string) string {
+	return pathWithNamespace[:strings.Index(pathWithNamespace, "/")]
+}
```

Comment:
I think that this function should be unit tested, what will happen when username will be in different format that expected

Reply:
Done